### ISSUE TITLE
Reduce build overhead of debugLog

### ISSUE THREAD

**Issue by HadrienG2** [acts-project/acts#347]
The `debugLog()` mechanism that is used throughout the Propagator machinery causes a surprisingly large amount of compile-time overhead in Propagator-related examples and unit tests.

Here are the reasons why that is the case, in order of decreasing importance:

- [ ] It unconditionally requires an `std::function<std::string()>` as its logging statement.
- [ ] Its definition is copy-pasted with relatively minor changes in many different places.
- [ ] It is a templated function within a templated type, which is probably a lot more templated than it needs to be.
- [ ] It internally uses boost's string processing routines, for arguably trivial purposes.

---

Requiring an `std::function` is the worst problem, because it forces a caller who wants to emit a trivial logging statement (which actually encompasses the vast majority of `debugLog` uses) to do something like this:

```c++
debugLog(state, [&] { return std::string("I AM A LOG"); });
```

Upon encountering this construct, a C++ compiler must...

- Create an anonymous lambda struct, with an `operator()` that contains the input code.
    * ...which in turn calls the constructor of `std::string` from a `const char*`, but I'll assume that is already instantiated somewhere.
- Instantiate the `std::function<std::string()>` constructor with that lambda as a parameter.
    * ...which among other things leads to the instantiation of an inner struct template from the std::function implementation that does nothing but forward calls to the lambda's `operator()` and forward destructor to the lambda's destructor, because that's how type erasure works.

Instead, it should be possible to pass a simple `const char*` to debugLog, like so:

```c++
debugLog(state, "I AM A LOG");
```

Even if that does nothing but call the following newly created trivial `debugLog` overload...

```c++
  template <typename propagator_state_t>
  void debugLog(propagator_state_t& state, const char* msg) const {
    debugLog(state, [&] { return std::string(msg); });
  }
```

...it will already have divided the aforementioned overhead by factors, because a single lambda struct will be reused for each `const char*`-based call to `debugLog`, given a certain enclosing type and debugLog "state" template parameter.

---

I should not need to explain why code duplication is a dubious practice in general. But given the relative simplicity of this function, I can understand why it was preferred over the perceived cognitive complexity of extracting this simple function to a centralized place.

To that code readability consideration, I will reply that as far as build overhead is concerned, making N copies of a function means having the compiler process it N times instead of one, i.e. multiplying its build-time overhead by a factor of N. Compilers are not smart enough to deduplicate near-perfect human copies of code snippets.

---

Similarly, templated types and functions are, by nature, a code duplication machine. They should be avoided for functions which are "hot" at compile time and whose runtime performance is largely irrelevant, like those debug logging features. Classic OO programming techniques should perform just fine in this particular case, and give much less work to the compiler.

---

Finally, the use of boost's string-processing algorithms that is made in some `debugLog` implementations...

```c++
      std::vector<std::string> lines;
      std::string input = logAction();
      boost::split(lines, input, boost::is_any_of("\n"));
      for (const auto& line : lines) {
        std::stringstream dstream;
        dstream << ">>>" << std::setw(state.options.debugPfxWidth) << vName
                << " | ";
        dstream << std::setw(state.options.debugMsgWidth) << line << '\n';
        state.options.debugString += dstream.str();
      }
```

...seems entirely trivial, and not worthy of the slight build overhead of using Boost's string processing algorithms, which is...

- 0.5s of build time spent parsing Boost's header (doesn't matter so much, since it's paid only once, but still huge for a friggin' header)
- 0.1s of build time spent instantiating `boost::split` per version of `debugLog` that uses the algorithm (a number which, if the above advice regarding code duplication and templates is followed, should ultimately drop to 1 or 2).

...nor does it seem worthy of the cognitive overhead of asking the reader of the code to go hunting how this particular Boost library works, given how generally badly engineered and documented Boost libraries are.

That being said, given how quantitatively low this particular overhead is, I think it's okay for it to stay around, if the other problems are resolved.

**Comment by paulgessinger** [comment]
Th bit about boost split I'm responsible for. This is mostly for readability reasons, so I'd like to have something along these lines, but it doesn't necessarily have to be the boost implementation. The consideration here was kind of that it doesn't matter for runtime performance since if the debug log is enabled runtime performance doesn't matter anymore anyway. But yeah, for the compile time it's probably not great. 


On the other hand, I think I'd like to get rid of the entire debugLog mechanism in favor the regular logging infrastructure anyway. I don't think we use the original purpose this was designed for and the way it's used right now it actually makes debugging off of the log more difficult I believe. 

I have a branch lying around that does about half of the change away from debugLog, I could revive that of there's interest.

**Comment by HadrienG2** [comment]
I am very strongly in favor of replacing debugLog with regular logging. The "delayed logging" design of the debugLog infrastructure has caused me significant debugging pain ever since it was introduced:

- It breaks interoperability with any other form of immediate logging (regular Acts logger, print statements...) because the output messages are emitted out of order, which makes interpretation of temporal execution almost impossible.
- It breaks any debugging tool based on the notion of stopping or diverting program execution, including of course debuggers.
- It means that if the component of interest hangs for whatever reason (e.g. Propagator tends to be easy to disturb in that way), no logging message will be emitted.

Back in the day, I tried to protest against it, but didn't manage to convince that it hurt a lot more than it helped. Maybe the build overhead argument will be the extra drop in the bucket that will do the trick...

**Comment by paulgessinger** [comment]
Yeah...

I think it was worth a try but we didn't end up reaping the benefits so I think we don't need to keep it around. 

I'll try to revive the branch I had next week.

**Comment by paulgessinger** [comment]
This has been addressed in #348.
